File: src/slang/parse/Stmt.java
package slang.parse;

import java.util.List;
import slang.lex.*;
import slang.parse.*;
import slang.runtime.*;

public abstract class Stmt {
  public interface Visitor<R> {
    R visitExpressionStmt(Expression stmt);
    R visitLetStmt(Let stmt);
    R visitPrintStmt(Print stmt);
    R visitMatchStmt(Match stmt);
  }
  public static class Expression extends Stmt {
    public Expression(Expr expr) {
      this.expr = expr;
    }

    public <R> R accept(Visitor<R> visitor) {
      return visitor.visitExpressionStmt(this);
    }

    public final Expr expr;
  }
  public static class Let extends Stmt {
    public Let(Pattern pattern, Expr init) {
      this.pattern = pattern;
      this.init = init;
    }

    public <R> R accept(Visitor<R> visitor) {
      return visitor.visitLetStmt(this);
    }

    public final Pattern pattern;
    public final Expr init;
  }
  public static class Print extends Stmt {
    public Print(Expr expr) {
      this.expr = expr;
    }

    public <R> R accept(Visitor<R> visitor) {
      return visitor.visitPrintStmt(this);
    }

    public final Expr expr;
  }
  public static class Match extends Stmt {
    public Match(List<Pattern> patterns, Expr expr) {
      this.patterns = patterns;
      this.expr = expr;
    }

    public <R> R accept(Visitor<R> visitor) {
      return visitor.visitMatchStmt(this);
    }

    public final List<Pattern> patterns;
    public final Expr expr;
  }

  public abstract <R> R accept(Visitor<R> visitor);
}

File: src/slang/parse/Stmt.scala
package slang.parse

import slang.runtime.Value
import slang.util.TryObject._

sealed trait Stmt

object Stmt {

  case class Expression(expr: Expr) extends Stmt

  case class Let(pattern: Pattern, init: Expr) extends Stmt

  case class Print(expr: Expr) extends Stmt

  case class Match(patterns: List[Pattern], expr: Expr) extends Stmt {
    def hasHead(pred: Pattern => Boolean): Boolean = pred(patterns.head)
    def isElementalHead: Boolean = hasHead(h => h.tryAsInstanceOf[Pattern.Literal].exists(_.value.isElemental))
    def isLiteralValueHead(value: Value): Boolean = hasHead(h => h.tryAsInstanceOf[Pattern.Literal].exists(_.value == value))
    def isLiteralTypeHead(value: Value): Boolean = hasHead(h => h.tryAsInstanceOf[Pattern.Literal].exists(_.value.getType == value.getType))
    def isVariableHead: Boolean = hasHead(h => h.isInstanceOf[Pattern.Id] || h.isInstanceOf[Pattern.Ignore])
    def isListHead: Boolean = hasHead(h => h.isInstanceOf[Pattern.SlangList])
    def tailMatch: Match = Match(patterns.tail, expr)
  }

}
